README
This is GDGraph, a package to generate charts, using Lincoln Stein's
GD.pm. See the documentation for some history and more information.

The latest release is available from CPAN: http://www.cpan.org/.

PLEASE NOTE: The release which includes this file is a set of bug-fixes and
feature additions on top of version 1.43, provided by various Perl users,
and assembled and and packaged up in the hope that they will be useful.
It should probably be considered a developer release, though there are no 
known regressions in it: a more official release may be forthcoming,
especially if that statement holds up.

INSTALLATION

Requirements:

	perl 5.6.0
	GD >= 1.19 (recommended >= 1.23)
	GD::Text::Align (part of the GDTextUtils package)

GD::Graph will not work with Perl 5.004. It most likely works with
5.005 (except for some odd problems with GD::Graph::Data::read(),
which seem to be related to some obscure 5.005_03 bug, which I have
not been able to track down), but I will no longer actively support it.

to install:

	perl Makefile.PL
	make
	make install

The tests are of limited value. The samples directory contains
examples which also function as a test suite. To see the samples:

	make samples
	(use viewer to see samples/sample*.png or .gif)

or

	cd samples
	make
	(use viewer to see sample*.png)

The documentation available is contained in GD::Graph.pm, as pod. See it
with 

perldoc GD::Graph

after installation or 

perldoc -F Graph.pm

before installation.

TESTED PLATFORMS

I have discontinued this list, because it is too hard to maintain.
There is hardly any code in the module that could provide problems
with inter-platform portability, but you never know. I will always
test the modules against the latest stable perl (5.8.0 right now) and
at least one older version (5.6.1). I will not explicitly test against
older versions of Perl. I normally test against the newest version of
GD, the last pre-2.0 version (1.41), and one version of the pre-PNG
days (1.19). All of this happens on Linux.

AUTHOR

Martien Verbruggen
mgjv@tradingpost.com.au

FIRST PNG PORT as Chart::PNGgraph

Steve Bonds
sbonds@agora.rdrop.com

MOST IMPORTANT CHANGES SINCE VERSION 1.34

See file CHANGES

MOST IMPORTANT CHANGES FOR VERSION 1.34

- Added GD::Graph::hbars, to create horizontal bar charts. Many thanks
  to brian d foy for doing almost all of the actual work. :)

MOST IMPORTANT CHANGES FOR VERSION 1.30

- Added GD::Graph::Data as the preferred way of working with data sets.
  Internally all code has been changed to use these objects.
- Many, many fundamental code changes, mainly to facilitate a future
  breakup of the code into separate Data, Chart and Graphics handling.
  All of this is internal, and there are not that many new options.
  (check the CHANGES file for more information)

MOST IMPORTANT CHANGES FOR VERSION 1.20

- Namespace change to GD::Graph
- No more links to GIF or PNG. We return a GD object now. It's the
  user's responsibility to save the right format. GIFgraph and
  (hopefully) Chart::PNGgraph will become wrappers around all this
- Font support no longer part of these modules. Instead the package
  GDTextUtils is necessary.
- Some bug fixes and extra options.

MOST IMPORTANT CHANGES SINCE VERSION 1.10

- Namespace change from GIFgraph --> Chart::PNGgraph
- No longer generates GIF format images due to Unisys LZW patent
- Supports TrueType fonts

MOST IMPORTANT CHANGES SINCE VERSION 1.04

- added bar_spacing option
- added limited support for numerical X axis for equidistant X values
  (NOT for a real numerical graph!), thanks to Scott Prahl.
- routines to automatically 'guess' nice upper and lower values for
  graphs have been replaced by one written by Scott Prahl

MOST IMPORTANT CHANGES SINCE VERSION 1.03

- Added mixed type graph.
  GIFgraph::mixed
  options types and default_type
- Added callback functionality to format y values
  option y_number_format
- Added control over axis label placement
  option [xy]_label_position

MOST IMPORTANT CHANGES SINCE VERSION 1.01

- Vertical printing of X axis labels
- CHANGED DEFAULTS for zero_axis and zero_axis_only
- Fixed problems with negative values, data ranges without 0, and
  printing of the x axis labels (I hope)
- No more support for perl 5.003. Since perl 5.005 is out, and I don't
  have a 5.003 here to test all the time, I won't support it anymore.

MOST IMPORTANT CHANGES SINCE VERSION 0.91

- Legends
- Support for negative numbers
- Ticks for X axis
- line types and widths
- Allow entries in data set to be undef (to skip them)
- Changed 'zero' angle for pie charts (was right, now is front/bottom)
- Definition of colours now possible. See PNGgraph::colour::read_rgb
  for details
- See file CHANGES for more

BUG FIXES IN THIS VERSION

See file CHANGES

REPORTING BUGS

Please use rt.cpan.org to report any bugs with GD::Graph or GD::Text.
Alternatively, you can email me, but I prefer it if you use the RT
interface. When reporting a bug, please include the version of the
module distribution you use, as well as your Perl version and platform
(perl -V output will do).

NOTES FOR WIN32 and Mac (pre OSX)

These modules _should_ work on all platforms.

Please direct any questions about the availability of these modules as
PPM packages to ActiveState. I only make this available as a standard
tar on CPAN. I have emailed ActiveState to ask what I was supposed to
do, and have, until today, not received any answers. 

THANKS

First of all. many, many thanks to Steve Bonds for adapting GIFgraph to
version 1.20 of GD (the first version that did not support GIF anymore),
and releasing it as Chart::PNGgraph. I really didn't have the time to
look at it, and he stepped in and fixed the headache that Unisys had
given me.

Others who have contributed are mentioned in the documentation for
GD::Graph in the AUTHOR section.

COPYRIGHT

GD::Graph
Copyright (c) 1999-2002 Martien Verbruggen. All rights reserved.
Chart::PNGgraph
Copyright (c) 1999 Steve Bonds.  All rights reserved.
GIFgraph
Copyright (c) 1995-1999 Martien Verbruggen. All rights reserved.

This package is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

I really felt like adding a clause to the copyright that excluded Unisys
and its employees from the use of any of my code, at least without a
license bought from me. I decided against it for several reasons, the
main one being that I would just feel too dirty doing that.



CHANGES
1.55 11 Jan 2023

- Fix failing XBM test resulting from some upstream changes
  Fixes: https://rt.cpan.org/Ticket/Display.html?id=140940
- Skip samples tests if libgd has image support disabled, which
  is the default starting with version 2.3.3. See:
  https://github.com/libgd/libgd/issues/428

1.54 21 Nov 2016

- Disable two Y axes alignment when any y[12]_{min,max}_value is defined
  RT#62665

1.53 08 Jul 2016

- Fix 'Illegal division by zero' when x_min_value and x_max_value
  are defined and x_tick_number set to 'auto' RT#73185
  Thanks to Bob Rogers, https://github.com/ruz/GDGraph/pull/12

1.52 28 Jan 2016

- y1_min_range and y2_min_range instead of min_range_1 and min_range_2,
  niether were documented before.
- Update documentation in regards to all *_min_range options available.

1.51 27 Dec 2015

- fix shadows rendering on cumulative bar charts
  thanks to https://github.com/Tordek
  see https://github.com/ruz/GDGraph/pull/4

1.50 27 Dec 2015

- run samples as part of test suite to make sure no sample crashes
  thanks to https://github.com/tynovsky
- properly define test requirements using newer MakeMaker

1.49 11 Mar 2015

- fix to Z-axis color filling in 3D pie charts (Debian Bug #489184)
- bump ExtUtils::MakeMaker dependency
- tiny improvement in the code of the samples

1.48 02 Aug 2013

- no code changes, just release enginering cleanup
- adjust MANIFEST.SKIP file so MANIFEST can be generated
  once again
- ship sample58.pl file, so `make samples` stop failing
- mention the current and past maintainers in META files
  as authors
- use newer CPAN::Meta and ExtUtils::MakeMaker, older
  versions generated META files without runtime prerequisites

1.47 28 Jun 2013

- experimental hide_overlapping_values option for bar graphs

1.46 26 Jun 2013

- This release is based on old work by Martien that was sitting
  in his repo
- x_last_label_skip option
- new samples and tweaks to old

1.45 21 Jun 2013

- read DISTRIBUTION STATUS in perldoc GD::Graph
- no code changes since 1.44

25 Apr 2007

- Patched bugs 21610, 20792, 20802, 23755 and 22932
- Updated POD to clarify current maintenance status, and encourage 
	bug reporting via RT (and to point out some external help resources)
- Release 1.44

17 May 2006

- Patched bugs 7307, 2944 (extended the fix to mixed graphs), 4177, and 16863.
- Fixed a fencepost error in pie.pm that caused an occasional segfault
	(reported by John Westlund and Hank Leininger)
- Fixed a bug that broke bar charts with no visible 0-axis 
	(reported by Steve Foster)

19 Feb 2006

- Patched bugs 2218, 4632 (which had two duplicates), 7881, and 15374
- Failed to update CHANGES file before releasing 1.4307 to CPAN (oops)

4 Feb 2006

- Patched bugs 16880 and 16791

19 Dec 2005

- Resolved the following bug reports/feature requests from RT:
	1363, 2944, 3346, 3850, 4104, 4380, 4384, 
	4469, 5275, 5282, 6751, 6786, 7287, 7819
- Straightened out $my_graph/$graph confusion in docs
- Added five new sample scripts
- Individual module versions now reflect a branch from original CVS
  versioning, in case things need to be merged at some later date.
- Uploading to CPAN as 1.4305 under ~BWARFIELD because I can't seem
  to get in touch with Martien Verbruggen.

1 July 2003

- Fixed yet another division by 0 problem, for two_axes. 
- Added more tests to axestype.t
- Cleaned up other test files.
- Release 1.43

19 June 2003

- Fixed another division by 0 problem, introduced in 1.41
- Added test t/axestype.pm, which now tests for division by 0 error.
- Released 1.42

17 June 2003

- Removed file BUGS from distribution. Too much work to keep up to
  date. Use rt.cpan.org, or email.
- Fixed skip() calls in tests to work with ancient versions of
  Test.pm.
- Made GD::Graph::Data::read() work with file handles under Perl
  5.005.
- Released 1.41

16 June 2003

- Fixed when zero axis inclusion is done for bar and are charts.
- Fixed code to reserve area for hbar charts last y axis label

11 June 2003

- Added no_axes option, changed sample56 to reflect this

30 May 2003

- Added version number for GD::Text PREREQ_PM
- Allow GD::Graph::Data::read() to read from file handle, instead of
  file
- Added tests for data file reading, and test data
- Release version 1.40

24 Feb 2003

- How come I never noticed this before? Right axis was disappearing
  when r_margin was zero.

22 Feb 2003

- Added patch by Ben Tilly from RT ticket 203 (manually, and much too
  late) to fix problems with picking decent values for axes when
  two_axes set to true. Added sample 57

20 Feb 2003

- Removed cetus font, because of unknown copyright status.
- Release 1.39 (skip 1.38, internal release)

11 Feb 2003, continued after 1.37 release

- Fixed version numbering
- Added limited, preliminary get_feature_coordinates support.

11 Feb 2003

- Fixed =head1 in GD/Graph/FAQ.pod
- release 1.37

10 Feb 2003

- Added some tests, mainly to make the CPAN testers happy :) The best
  way to test is still to use the samples.

- Preparation for 1.36

Jan 2003

- Made detection of output formats more robust. Newer GDs break on
  simply testing with UNIVERSAL::can.
- Added some documentation on error handling.

12 Jun 2002 - 21 Jun 2002

- Fixed various bugs: 
  - Area charts don't allow undefined variables, die on hotspot code
  - allow "0.00" to be equal to 0 when determining min and max values
    for axes
  - fixed shadows for cumulative bars

- Preparation for release 1.35

9 Jun 2002

- I just realised this file is severely out of date. I'll only keep
  track of the really big changes here, since I really can't remember
  what I've fixed and changed sine 1.33, and it's really too much work
  to go through all CVS comments.
- Added hbars.pm, and put most of the framework in place to allow the
  other charts to be plotted rotated as well.
- preparation of release 1.34

7 Oct 2000

- Addition of undefined data set colour.
- Addition of bar_width option
- Preparation of release 1.33

May - Sep 2000


Makefile.PL
# This -*- perl -*- script makes the Makefile
# $Id: Makefile.PL,v 1.1 2005/12/14 04:21:45 ben Exp $
 
#--- Distribution section ---
 
use ExtUtils::MakeMaker 6.76;
use Config;
 
#use ExtUtils::Manifest;
#ExtUtils::Manifest::skipcheck();
 
my @clean = qw(*.old *.save tags);
push @clean, map { "*/$_" } @clean;

WriteMakefile(
        'VERSION_FROM'  => 'Graph.pm',  #finds $VERSION
        'DISTNAME'      => 'GDGraph',
        'NAME'          => 'GD::Graph',

        ($] >= 5.005 ?
            ('ABSTRACT'	    => 'Produces charts with GD',
             'AUTHOR'       => [
                 'Martien Verbruggen <mgjv@tradingpost.com.au>',
                 'Benjamin Warfield <bwarfield@cpan.org>',
                 'Ruslan Zakirov <Ruslan.Zakirov@gmail.com>',
                 'Best Practical Solutions <modules@bestpractical.com>',
            ]
            ):()
        ),

        BUILD_REQUIRES   => {
            'ExtUtils::MakeMaker' => '6.76',
        },
        CONFIGURE_REQUIRES   => {
            'ExtUtils::MakeMaker' => '6.76',
        },
        TEST_REQUIRES => {
            'Capture::Tiny'       => '0.30',
            'Test::Exception'     => '0.40',
            'FindBin'             => '0',
            'Test::More'          => '0.88',
        },
        test => {
            RECURSIVE_TEST_FILES => 1,
        },

        'PREREQ_PM'     => { 
            'GD'	    => '1.18', 
            'GD::Text'      => '0.80',
        },
 
        'dist'          => {
            'COMPRESS'      => 'gzip -9f',
            'SUFFIX'        => 'gz', 
        },

        LICENSE         => 'perl',

        META_MERGE      => {
            'meta-spec' => { version => 2 },

            resources => {
                bugtracker => {
                    web => 'http://rt.cpan.org/Public/Dist/Display.html?Name=GDGraph',
                    mailto => 'bug-GDGraph@rt.cpan.org',
                },
                repository => {
                    type => 'git',
                    url  => 'git://github.com/bestpractical/gdgraph.git',
                    web  => 'https://github.com/bestpractical/gdgraph',
                },
            },

        },

        'clean'         => { 'FILES' => join(" ", @clean), },
);

sub MY::postamble 
{
qq(
samples: all
	\@(cd samples && make PERL=$Config{'perlpath'})
	\@echo Samples have been created in samples directory

tags:
	ptags lib/GD/*.pm lib/GD/Graph/*.pm
)
}

sub MY::libscan
{
    my ($self, $path) = @_;
    return if
	$path =~ m:(^|/)\..+\.swp$:	    or
	$path =~ m:(^|/)\.#.+\d$:	    or
	$path =~ m:\b(RCS|CVS|SCCS)\b:	    ;
    return $path;
}

print <<END;

The automatic tests for GDGraph are not really a solid workout of the
library. The best way to test the package is to run the examples
before installing it.  You can run the examples in the samples
directory with `make samples` or by going into that directory, and
just running `make`.
If that fails, please read samples/Makefile.

UPDATE 1: Running the examples is part of the test procedure now. You
are still encouraged to run check them out to learn from them and
to see the ouput images.

UPDATE 2: The samples fail if libgd is installed without image support
enabled, which is the default starting with version 2.3.3. The samples
should be automatically skipped if this is detected. To run the tests,
install libgd with image support enabled.
END

